Minigrep.hs

```diff
@@ -0,0 +1,86 @@
+import Control.Applicative (Alternative ((<|>)))
+import Control.Monad (liftM)
+import Control.Monad.Trans.Maybe (MaybeT (MaybeT, runMaybeT))
+import Data.Char (toLower)
+import Data.List (isInfixOf)
+import Data.Maybe (fromMaybe, isJust)
+import System.Directory (doesFileExist)
+import System.Environment (getArgs, lookupEnv)
+
+type Query = String
+
+type FileContents = String
+
+data Config = Config {query :: Query, filePath :: FilePath, caseSensitive :: Bool}
+
+lowerCase :: String -> String
+lowerCase = map toLower
+
+isCaseSensitive :: IO Bool
+isCaseSensitive = do
+  envVariable <- lookupEnv "CASE_SENSITIVE"
+  case envVariable of
+    Nothing -> return False
+    Just caseSensitive -> return $ lowerCase caseSensitive == "true"
+
+type Args = [String]
+
+getQuery :: Args -> Maybe Query
+getQuery args =
+  if null args
+    then Nothing
+    else Just $ head args
+
+getFilePath :: Args -> Maybe FilePath
+getFilePath args =
+  if length args < 2
+    then Nothing
+    else Just $ args !! 1
+
+getConfig :: MaybeT IO Config
+getConfig =
+  do
+    args <- MaybeT $ Just <$> getArgs
+
+    query <- (MaybeT . return . getQuery $ args) <|> MaybeT (Nothing <$ putStrLn "Didn't get a query!")
+    filePath <- (MaybeT . return . getFilePath $ args) <|> MaybeT (Nothing <$ putStrLn "Didn't get a file path!")
+
+    caseSensitive <- MaybeT $ Just <$> isCaseSensitive
+    return $ Config {query = query, filePath = filePath, caseSensitive = caseSensitive}
```

Comment:
In general we prefer `pure` over `return` (not a historic mistake*, shorter, less confusing to beginners).

There's quite a lot of duplication here, easily avoided by introducing the right (local) helpers, i.e.
```haskell
-- not sure about MaybeT's type here, but you get the idea
query <- handleMissing getQuery "Didn't get a query!"
path <- handleMissing getFilePath "Didn't get a file path!"
    where
        handleMissing :: Monad m => ([a] -> Maybe a) -> String -> MaybeT m String
        handleMissing extract message =
            MaybeT $ (pure $ extract args) <|> (Nothing <$ putStrLn message)
```

*the only reason `return` exists is because till recently `Applicative` was not a super class of `Monad`... The standard library is full of such unfortunate mistakes ironed in due to backwards compatibility. However, `purescript`'s standard library fixes pretty much everything (and adds some improvements).